packages/playwright-core/src/server/screenshotter.ts

```diff
@@ -107,50 +116,92 @@ export class Screenshotter {
       return buffer;
     });
   }
 
-  private async _screenshot(progress: Progress, format: 'png' | 'jpeg', documentRect: types.Rect | undefined, viewportRect: types.Rect | undefined, fitsViewport: boolean | undefined, options: types.ElementScreenshotOptions): Promise<Buffer> {
-    if ((options as any).__testHookBeforeScreenshot)
-      await (options as any).__testHookBeforeScreenshot();
-    progress.throwIfAborted(); // Screenshotting is expensive - avoid extra work.
-    const shouldSetDefaultBackground = options.omitBackground && format === 'png';
-    if (shouldSetDefaultBackground) {
-      await this._page._delegate.setBackgroundColor({ r: 0, g: 0, b: 0, a: 0 });
-      progress.cleanupWhenAborted(() => this._page._delegate.setBackgroundColor());
-    }
-    progress.throwIfAborted(); // Avoid extra work.
-
-    const restoreBlinkingCaret = async () => {
-      await Promise.all(this._page.frames().map(async frame => {
-        frame.nonStallingEvaluateInExistingContext('window.__cleanupScreenshot && window.__cleanupScreenshot()', false, 'utility').catch(() => {});
-      }));
-    };
+  async _preparePageForScreenshot(progress: Progress, disableAnimations: boolean) {
     await Promise.all(this._page.frames().map(async frame => {
-      await frame.nonStallingEvaluateInExistingContext((function() {
+      await frame.nonStallingEvaluateInExistingContext('(' + (function(disableAnimations: boolean) {
         const styleTag = document.createElement('style');
         styleTag.textContent = `
-          * { caret-color: transparent !important; }
-          * > * { caret-color: transparent !important; }
-          * > * > * { caret-color: transparent !important; }
-          * > * > * > * { caret-color: transparent !important; }
+          *,
+          * > *,
+          * > * > *,
+          * > * > * > *,
           * > * > * > * > * { caret-color: transparent !important; }
         `;
         document.documentElement.append(styleTag);
+        const infiniteAnimationsToResume: Animation[] = [];
+        const cleanupCallbacks: (() => void)[] = [];
+
+        if (disableAnimations) {
+          const collectRoots = (root: Document | ShadowRoot, roots: (Document|ShadowRoot)[] = []): (Document|ShadowRoot)[] => {
+            roots.push(root);
+            const walker = document.createTreeWalker(root, NodeFilter.SHOW_ELEMENT);
+            do {
+              const node = walker.currentNode;
+              const shadowRoot = node instanceof Element ? node.shadowRoot : null;
+              if (shadowRoot)
+                collectRoots(shadowRoot, roots);
+            } while (walker.nextNode());
+            return roots;
+          };
+          const handleAnimations = (root: Document|ShadowRoot): void => {
+            for (const animation of root.getAnimations()) {
+              try {
+                animation.finish();
+              } catch (e) {
+                animation.cancel();
```

Comment:
This may [reject some promises](https://developer.mozilla.org/en-US/docs/Web/API/Animation/cancel#exceptions) that could be unexpected to the page. Any thoughts?

Reply:
Attempts to use the `animation.updatePlaybackRate(0);animation.currentTime = 0;` didn't work – no more ideas here.

Luckily, infinite animations are unlikely to have callbacks.